lib/active_support/cache/database_store.rb

```diff
@@ -74,10 +81,23 @@ def write_entry(key, entry, _options = nil)
         record = @model.where(key: key).first_or_initialize
         expires_at = Time.zone.at(entry.expires_at) if entry.expires_at
         record.update! value: Marshal.dump(entry.value), version: entry.version.presence, expires_at: expires_at
+      ensure
+        cleanup if @auto_cleanup || max_size_exceeded?
+        delete_oldest_entry if max_size_exceeded? # <- Only happens when running cleanup was not enough
```

Comment:
I can see this being a useful thing, but I am worried about the number of SQL queries that it requires (up to 6!). In the same vein as with #14 I wonder if this is better implemented as a timed option. Instead of running up-to 6 queries on every write, ensure there is at least e.g. 1 minute between each check, i.e. make this an "approximate" rather than an absolute thing. 

Furthermore, you could optimise here:
1. you could make `cleanup` return an integer (which I think it already does) to indicate the number of records removed, that would potentially save you a query
2. always calling `max_size_exceeded?` twice is not necessary anyway, e.g. if the first attempt already returns false